app/controllers/users/sessions_controller.rb

```diff
@@ -24,4 +29,69 @@ class Users::SessionsController < Devise::SessionsController
   # def configure_sign_in_params
   #   devise_parameter_sanitizer.permit(:sign_in, keys: [:attribute])
   # end
+  include ApplicationHelper
+  def authenticate_with_otp_two_factor
+    user = self.resource = find_user
+    if user_params[:otp_attempt].present? && session[:otp_user_id]
+      authenticate_user_with_otp_two_factor(user)
+    elsif user&.valid_password?(user_params[:password])
+     if user&.current_sign_in_ip != user&.last_sign_in_ip  
+      prompt_for_otp_two_factor(user)
+     else
+        sign_in(user)
+     end
+    end
+  end
+
+  private
+
+  def valid_otp_attempt?(user)
```

Comment:
@ajaystha42 these methods are copy pasted in multiple controllers, `DRY` this up

Reply:
`valid_otp_attempt` and `prompt_for_otp_two_factor` and `authenticate_user_with_otp_two_factor` and `otp_two_factor_enabled` and `enable_2fa?` need to defined as reusable functions.